kernel/cooking.ml

```diff
@@ -165,83 +215,93 @@ type 'opaque result = {
         C'{0, ..., n - 1, n, ..., n + m - 1} ]
-      together with the instance
-      [u₀ ... uₙ₋₁ Var(0) ... Var (m - 1)].
+      together with the substitution
+      [u₀ ↦ Var(0), ... ,uₙ₋₁ ↦ Var(n - 1), Var(0) ↦  Var(n), ..., Var(m - 1) ↦  Var (n + m - 1)].
   *)
-  if (Univ.Instance.is_empty subst) then
-    (** Still need to take the union for the constraints between globals *)
-    subst, (AbstractContext.union abs_ctx auctx)
+  let open Univ in
+  let n = AbstractContext.size abstr.abstr_uctx in
+  if (Int.equal n 0) then
+    (** Optimization: still need to take the union for the constraints between globals *)
+    { abstr with abstr_uctx = AbstractContext.union abstr.abstr_uctx auctx }
   else
-    let open Univ in
+    let subst = abstr.abstr_usubst in
     let ainst = make_abstract_instance auctx in
-    let subst = Instance.append subst ainst in
-    let substf = make_instance_subst subst in
+    let substf = Univ.lift_level_subst n (make_instance_subst ainst) in
+    let substf = Univ.merge_level_subst subst substf in
     let auctx = Univ.subst_univs_level_abstract_universe_context substf auctx in
-    subst, (AbstractContext.union abs_ctx auctx)
-
-let lift_univs subst auctx0 = function
+    let auctx' = AbstractContext.union abstr.abstr_uctx auctx in
+    { abstr with abstr_uctx = auctx'; abstr_usubst = substf }
+
+let lift_poly_univs info auctx =
+  (** The constant under consideration is quantified over a
+      universe context [auctx]; it has to be quantified further over
+      [abstr.abstr_uctx] leading to a new abstraction recipe valid
+      under the quantification; that is if we had a judgment
+      [G, Δ ⊢ ΠG'.J] to be turned, thanks to [abstr] =
+      [{ctx:=Δ;uctx:=G;subst:=σ;usubst:=τ}], into
+      [⊢ ΠG.ΠΔ.(ΠG'.J)[σ][τ])], we build the new
+      [{ctx:=Δ;uctx:=G(G'[ττ']);subst:=σ;usubst:=ττ'}], for some [τ']
+      built by [discharge_abstract_universe_context], that works on
+      [J], that is, that allows to turn [GG'[ττ'], Δ ⊢ J] into
+      [⊢ ΠG.ΠΔ.(ΠG'.J)[σ][τ]] via [⊢ ΠG(G'[ττ']).ΠΔ.(J[σ][ττ'])] *)
+  let abstr_info = discharge_abstract_universe_context info.abstr_info auctx in
+  { info with abstr_info }, abstr_info.abstr_uctx
+
+let lift_univs info = function
   | Monomorphic ctx ->
-    assert (AbstractContext.is_empty auctx0);
-    subst, (Monomorphic ctx)
+    assert (AbstractContext.is_empty info.abstr_info.abstr_uctx); (* No monorphic constants in a polymorphic section *)
+    info, Monomorphic ctx
   | Polymorphic auctx ->
-    let subst, auctx = discharge_abstract_universe_context subst auctx0 auctx in
-    subst, (Polymorphic auctx)
+    let info, auctx = lift_poly_univs info auctx in
+    info, Polymorphic auctx
+
+let subst_private_univs info = function
+  | Opaqueproof.PrivateMonomorphic () as priv ->
+    let () = assert (AbstractContext.is_empty info.abstr_info.abstr_uctx) in
+    let () = assert (is_empty_level_subst info.abstr_info.abstr_usubst) in
+    priv
```

Comment:
Can't there be private universes also for monomorphic constants?

Reply:
We don't pass them to cook_constr (see drop_mono in opaqueproof)

Reply:
We don't pass them to cook_constr (see drop_mono in opaqueproof)

Reply:
I don't understand well the story about `PrivateMonomorphic`. Is the apparentl loss of information in `drop_mono` to be related to the extraction of `delayed_cst` in `Safe_typing.add_constant`, so that the information is actually not lost?

Reply:
This data is redundant for the kernel, and only used for vio computation. I have a series of patches that modify this area of the code but I've not submitted them yet because they break weird invariants in Declare that I need to fix first.